streaming/base/shared.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2022 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Synchronization primitives that live in shared memory.
+
+For when using `threading` or `multiprocessing` from the python standard library won't do, because
+we are coordinating separately instantiated pytorch worker processes.
+"""
+
+import os
+from multiprocessing.shared_memory import SharedMemory
+from time import sleep
+
+import numpy as np
+from filelock import FileLock
+
+TICK = 0.007
+
+
+class SharedBarrier:
+    """A barrier that works inter-process using a file lock and shared memory.
+
+    Args:
+        count (int): Number of processes to synchronize.
+        filelock_path (str): Path to lock file on local filesystem.
+        shm_path (str): Shared memory object name in /dev/shm.
+    """
+
+    def __init__(self, count: int, filelock_path: str, shm_path: str) -> None:
+        self.count = count
+        self.filelock_path = filelock_path
+        self.shm_path = shm_path
+
+        dirname = os.path.dirname(filelock_path)
+        os.makedirs(dirname, exist_ok=True)
+        self.lock = FileLock(filelock_path)
+
+        size = 3 * np.int32().nbytes
+        try:
+            self._shm = SharedMemory(shm_path, True, size)
+        except:
+            self._shm = SharedMemory(shm_path)
+
+        self._arr = np.ndarray(3, buffer=self._shm.buf, dtype=np.int32)
+        self.num_enter = 0
+        self.num_exit = count
+        self.flag = True
+
+    def __del__(self):
+        """Destructor clears array that references shm."""
+        del self._arr
```

Comment:
why not calling `self._shm.close() ` and `self._shm.unlink()` in the body of destructor?

Reply:
because this barrier's shared memory is almost certainly attached to other processes here

numpy arrays whose buffers are backed by shm don't reference count properly

Reply:
Hmmm I see. Then this could lead to a memory leak right ?

Reply:
no, shms are garbage collected when the last holder goes out of scope, also i manually delete the numpy array there